### PR TITLE
Remove some `DataErrorKind` variants

### PR DESCRIPTION
**Pull request by robertbastian**

### CODE REVIEW

File: provider/core/src/response.rs [hunk 1]
```diff
@@ -386,7 +386,7 @@ where
             DataPayloadInner::Yoke(yoke) => yoke.try_into_yokeable().ok(),
             DataPayloadInner::StaticRef(_) => None,
         }
-        .ok_or(DataErrorKind::InvalidState.with_str_context("try_unwrap_owned"))
+        .ok_or(DataError::custom("DataPayload not owned"))
```
[the remark this answers is not in the document]
Comment (robertbastian) [reply provider/core/src/response.rs:389]: This will duplicate the error in the two call sites in `any.rs`

**Review by sffc — CHANGES_REQUESTED** [review]
This PR gets rid of two errors:

- KeyLocaleSyntax: replaces it with locale parse error. Seems good.
- InvalidState: replaces it with DataError::custom. Not great; we should avoid DataError::custom coming from the core crate.

It would be nice to get rid of InvalidState via some other mechanism, but replacing it with DataError::custom seems like a regression.

**Comment by robertbastian** [comment]
Well, it gets rid of the one usage of `InvalidState`, which requires a custom error message to express why the state is invalid. An error is always some kind of invalid state of the program, so I don't consider this variant useful.

**Comment by robertbastian** [comment]
Note that deserialization, in the core crate, also returns custom errors. Downcasting Any is not that different.

**Review by sffc — APPROVED** [review]

**Pull request merged**